src/Database/LSMTree/Internal/RefCount.hs

```diff
@@ -0,0 +1,115 @@
+-- count reaches @0@, the finaliser will be run.
+data RefCounter m = RefCounter {
+    countVar     :: !(MutVar (PrimState m) RefCount)
+  , finaliserVar :: !(MutVar (PrimState m) (Maybe (m ())))
+  }
+
+instance NFData (RefCounter m) where
+  rnf RefCounter{countVar, finaliserVar} =
+      rwhnf countVar `seq` rwhnf finaliserVar
+
+{-# SPECIALISE unsafeMkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Like 'mkRefCounterN', but throws an error if the initial @n < 1@
+unsafeMkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (RefCounter m)
+unsafeMkRefCounterN !n finaliser = mkRefCounterN n finaliser >>= \case
+    Nothing -> error "unsafeMkRefCounterN: n < 1"
+    Just rc -> pure rc
+
+{-# SPECIALISE mkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (Maybe (RefCounter IO)) #-}
+-- | Make a reference counter with initial value @n@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+--
+-- Returns 'Nothing' if @n < 1@, and 'Just' otherwise.
+mkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (Maybe (RefCounter m))
+mkRefCounterN !n finaliser
+  | n < RefCount 1 = pure Nothing
+  | otherwise = do
+    -- evaluate the finaliser a little bit before we store it
+    let !() = case finaliser of
+                Nothing       -> ()
+                Just !_action -> ()
+    countVar <- newMutVar n
+    finaliserVar <- newMutVar finaliser
+    pure $! Just $! RefCounter{countVar, finaliserVar}
+
+{-# SPECIALISE mkRefCounter1 :: Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Make a reference counter with initial value @1@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+mkRefCounter1 :: PrimMonad m => Maybe (m ()) -> m (RefCounter m)
+mkRefCounter1 finaliser = fromJust <$> mkRefCounterN (RefCount 1) finaliser
+
+data RefCountError = RefCountAlreadyZero
+  deriving stock (Show, Eq)
+  deriving anyclass Exception
+
+{-# SPECIALISE addReference :: RefCounter IO -> IO () #-}
+-- | Increase the reference counter by one.
+--
+-- If the reference counter was already at @0@, a 'RefCountError' error is
+-- thrown.
+addReference :: (PrimMonad m, MonadThrow m) => RefCounter m -> m ()
+addReference RefCounter{countVar} = do
+    success <- atomicModifyMutVar' countVar $ \c ->
+      if c == RefCount 0 then (c, False)
+                         else (incRefCount c, True)
+    unless success $ throwIO RefCountAlreadyZero
+
+{-# SPECIALISE readRefCount :: RefCounter IO -> IO RefCount #-}
+readRefCount :: PrimMonad m => RefCounter m -> m RefCount
+readRefCount RefCounter{countVar} = readMutVar countVar
+
+{-# SPECIALISE removeReference :: RefCounter IO -> IO () #-}
+-- | Decrease the reference counter by one.
+--
+-- When the reference counter reaches zero, the finaliser is run promptly.
+--
+-- If the reference counter was already at @0@, then this function is
+-- idempotent. The finaliser will not run again.
```

Comment:
But it should throw an exception! It means someone got their counting wrong if they're trying to release a reference on something that already had its last referenced removed.

Reply:
Hmm, yes I'm not so sure I remember why I made it idempotent... I'll try to retrace my steps

Reply:
Doesn't look like it's necessary for it to be idempotent